src/libre/re_char_class.c

```diff
@@ -0,0 +1,802 @@
+		pp_iter(f, n->u.subtract.mask, indent + 4);
+		break;
+	default:
+		fprintf(stderr, "(MATCH FAIL)\n");
+		assert(0);
+	}
+}
+
+void
+re_char_class_ast_prettyprint(FILE *f,
+    struct re_char_class_ast *ast, size_t indent)
+{
+	pp_iter(f, ast, indent);
+}
+
+static void
+free_iter(struct re_char_class_ast *n)
+{
+	assert(n != NULL);
+	switch (n->t) {
+	case RE_CHAR_CLASS_AST_CONCAT:
+		free_iter(n->u.concat.l);
+		free_iter(n->u.concat.r);
+		break;
+	case RE_CHAR_CLASS_AST_SUBTRACT:
+		free_iter(n->u.subtract.ast);
+		free_iter(n->u.subtract.mask);
+		break;
+	case RE_CHAR_CLASS_AST_LITERAL:
+	case RE_CHAR_CLASS_AST_RANGE:
+	case RE_CHAR_CLASS_AST_NAMED:
+	case RE_CHAR_CLASS_AST_CHAR_TYPE:
+	case RE_CHAR_CLASS_AST_FLAGS:
+		break;
+
+	default:
+		fprintf(stderr, "(MATCH FAIL)\n");
+		assert(0);
+	}
+	free(n);	
+}
+
+void
+re_char_class_ast_free(struct re_char_class_ast *ast)
+{
+	free_iter(ast);
+}
+
+void
+re_char_class_free(struct re_char_class *cc)
+{
+	free(cc);
+}
+
+static int
+comp_iter(struct re_char_class *cc, struct re_char_class_ast *n)
+{
+	assert(cc != NULL);
+	assert(n != NULL);
+	switch (n->t) {
+	case RE_CHAR_CLASS_AST_CONCAT:
+		if (!comp_iter(cc, n->u.concat.l)) { return 0; }
+		if (!comp_iter(cc, n->u.concat.r)) { return 0; }
+		break;
+	case RE_CHAR_CLASS_AST_LITERAL:
+		cc_add_byte(cc, n->u.literal.c);
+		break;
+	case RE_CHAR_CLASS_AST_RANGE:
+		cc_add_range(cc, &n->u.range.from, &n->u.range.to);
+		break;
+	case RE_CHAR_CLASS_AST_NAMED:
+		if (!cc_add_named_class(cc, n->u.named.id)) { return 0; }
+		break;
+	case RE_CHAR_CLASS_AST_CHAR_TYPE:
+		if (!cc_add_char_type(cc, n->u.char_type.id)) { return 0; }
+		break;
+	case RE_CHAR_CLASS_AST_FLAGS:
+		cc->flags |= n->u.flags.f;
+		break;
+	case RE_CHAR_CLASS_AST_SUBTRACT:
+	{
+		struct re_char_class *mask;
+		if (!comp_iter(cc, n->u.subtract.ast)) { return 0; }
+
+		mask = re_char_class_ast_compile(n->u.subtract.mask);
+		if (mask == NULL) { return 0; }
+		cc_mask(cc, mask);
+		re_char_class_free(mask);
+		break;
+	}
+	default:
+		fprintf(stderr, "(MATCH FAIL)\n");
+		assert(0);
+	}
+
+	return 1;
+}
+
+struct re_char_class *
+re_char_class_ast_compile(struct re_char_class_ast *cca)
+{
+	struct re_char_class *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return NULL; }
+
+	if (!comp_iter(res, cca)) {
+		cc_free(res);
+		return NULL;
+	}
+
+	if (res->flags & RE_CHAR_CLASS_FLAG_MINUS) {
+		cc_add_byte(res, '-');
+		res->flags &=~ RE_CHAR_CLASS_FLAG_MINUS;
+	}
+
+	if (res->flags & RE_CHAR_CLASS_FLAG_INVERTED) {
+		cc_invert(res);
+		res->flags &=~ RE_CHAR_CLASS_FLAG_INVERTED;
+	}
+
+	return res;
+}
+
+struct re_char_class *
+re_char_class_type_compile(enum ast_char_type_id id)
+{
+	struct re_char_class *res = calloc(1, sizeof(*res));
+	if (res == NULL) { return NULL; }
+
+	cc_add_char_type(res, id);
+	
+	return res;
+}
+
+
+static void
+bitset_pos(unsigned char byte, unsigned *pos, unsigned char *bit)
+{
+	*pos = byte / 8;
+	*bit = 1U << (byte & 0x07);
+}
```

Comment:
I forgot to say, there's a bitmap under `<adt/bitmap.h>` already - which the character classes did originally use before converting them to graft on pre-constructed DFA as part of the unicode support.